cmd/legalhold-main.go

```diff
@@ -0,0 +1,176 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/fatih/color"
+	"github.com/minio/cli"
+	"github.com/minio/mc/pkg/probe"
+	minio "github.com/minio/minio-go/v6"
+	"github.com/minio/minio/pkg/console"
+)
+
+var (
+	lhFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "recursive, r",
+			Usage: "apply legal hold recursively",
+		},
+	}
+)
+var legalHoldCmd = cli.Command{
+	Name:   "legalhold",
+	Usage:  "set object legal hold for objects with a given prefix",
+	Action: mainLegalHold,
+	Before: setGlobalsFromContext,
+	Flags:  append(lhFlags, globalFlags...),
+	CustomHelpTemplate: `NAME:
+   {{.HelpName}} - {{.Usage}}
+ 
+ USAGE:
+   {{.HelpName}} [FLAGS] TARGET [ON | OFF]
+ 
+ FLAGS:
+   {{range .VisibleFlags}}{{.}}
+   {{end}}
+ 
+ EXAMPLES:
+	1. Enable object legal hold for objects in a given prefix
+	  $ {{.HelpName}} myminio/mybucket/prefix ON --recursive
+
+	2. Enable legal hold on a specific object
+	  $ {{.HelpName}} myminio/mybucket/prefix/obj.csv ON
+ `,
+}
+
+// Structured message depending on the type of console.
+type legalHoldCmdMessage struct {
+	LegalHold minio.LegalHoldStatus `json:"legalhold"`
+	URLPath   string                `json:"urlpath"`
+	Status    string                `json:"status"`
+	Err       error                 `json:"error"`
+}
+
+// Colorized message for console printing.
+func (l legalHoldCmdMessage) String() string {
+	if l.Err != nil {
+		return console.Colorize("LegalHoldMessageFailure", "Cannot set object legal hold status `"+l.URLPath+"`."+l.Err.Error())
+	}
+	return ""
+}
+
+// JSON'ified message for scripting.
+func (l legalHoldCmdMessage) JSON() string {
+	msgBytes, e := json.MarshalIndent(l, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+	return string(msgBytes)
+}
+
+// setRetention - Set Retention for all objects within a given prefix.
+func setLegalHold(urlStr string, lhold *minio.LegalHoldStatus, isRecursive bool, encKeyDB map[string][]prefixSSEPair) error {
+	clnt, err := newClient(urlStr)
+	if err != nil {
+		fatalIf(err.Trace(), "Cannot parse the provided url.")
+	}
+
+	// Quit early if urlStr does not point to an S3 server
+	switch clnt.(type) {
+	case *fsClient:
+		fatal(errDummy().Trace(), "Legal hold not supported for file system.")
+	}
+
+	alias, _, _ := mustExpandAlias(urlStr)
+	if !isRecursive {
+		sse := getSSE(urlStr, encKeyDB[alias])
```

Comment:
Why do we need `sse` to set `legalhold` ?

Reply:
removed - stat doesn't add much value, so letting list fail if recursive flag is not passed